vendor/src/github.com/docker/libnetwork/drivers/overlay/filter.go

```diff
@@ -12,6 +12,13 @@ const globalChain = "DOCKER-OVERLAY"
 
 var filterOnce sync.Once
 
+var filterChan = make(chan struct{}, 1)
+
+func filterWait() func() {
```

Comment:
IMO it has the same chance of a dealock and race as a mutex because there is no way to try or cancel. This is basically non-obvious, slow mutex.
However, it's probably up to @mrjana to decide what to use in libnetwork :)